File: goopang-java/src/main/java/com/example/goopangjava/service/CartService.java
package com.example.goopangjava.service;

import com.example.goopangjava.entity.cart.Cart;
import com.example.goopangjava.entity.cart.CartItem;
import com.example.goopangjava.entity.user.User;
import com.example.goopangjava.input.AddCartItemInput;
import com.example.goopangjava.input.DeleteCartItemInput;
import com.example.goopangjava.repository.Database;
import org.apache.coyote.BadRequestException;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class CartService {
    private final ProductService productService;

    public CartService(ProductService productService) {
        this.productService = productService;
    }

    public Cart getUserCart(String userId) throws BadRequestException {
        return Database.getInstance().carts.stream()
                .filter(cart -> cart.getUser().getId().equals(userId))
                .findFirst()
                .map(cart -> {
                    cart.setItems(Database.getInstance().cartItems.stream()
                            .filter(cartItem -> cartItem.getCart().getId().equals(cart.getId()))
                            .toList());
                    cart.setTotalAmount(cart.getItems().stream()
                            .mapToDouble(item -> item.getProduct().getPrice() * item.getQuantity())
                            .sum());
                    return cart;
                })
                .orElseThrow(() -> new BadRequestException("Cart not found"));
    }

    public Cart addUserCart(User user) {
        Cart cart = new Cart(
                UUID.randomUUID().toString().substring(0, 5),
                user
        );
        Database.getInstance().carts.add(cart);
        return cart;
    }

    public Cart addCartItem(AddCartItemInput addCartItemInput) throws BadRequestException {
        CartItem cartItem = new CartItem(
                UUID.randomUUID().toString().substring(0, 5),
                addCartItemInput.getQuantity(),
                productService.getProduct(addCartItemInput.getProductId()),
                getUserCart(addCartItemInput.getUserId())
        );
        Database.getInstance().cartItems.add(cartItem);
        return getUserCart(addCartItemInput.getUserId());
    }

    public Cart deleteCartItem(DeleteCartItemInput deleteCartItemInput) throws BadRequestException {
        Database.getInstance().cartItems.removeIf(cartItem -> cartItem.getId().equals(deleteCartItemInput.getCartItemId()));
        return getUserCart(deleteCartItemInput.getUserId());
    }
}

File: goopang-kotlin/src/main/kotlin/com/example/goopang/service/CartService.kt
package com.example.goopang.service

import com.example.goopang.entity.cart.Cart
import com.example.goopang.entity.cart.CartItem
import com.example.goopang.entity.user.User
import com.example.goopang.input.AddCartItemInput
import com.example.goopang.input.DeleteCartItemInput
import com.example.goopang.repository.Database
import org.apache.coyote.BadRequestException
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class CartService(
    private val productService: ProductService
) {

    fun getUserCart(userId: String): Cart {
        return Database.carts.firstOrNull { it.user.id == userId }
            ?.also { cart -> cart.items = Database.cartItems.filter { it.cart.id == cart.id } }
            ?.also { cart -> cart.totalAmount = cart.items.sumOf { it.product.price * it.quantity } }
            ?: throw BadRequestException("Cart not found")
    }

    fun addUserCart(user: User): Cart {
        return Cart(
            id = UUID.randomUUID().toString().substring(0, 5),
            user = user
        )
            .also { cart -> Database.carts.add(cart) }
    }

    fun addCartItem(addCartItemInput: AddCartItemInput): Cart {
        CartItem(
            id = UUID.randomUUID().toString().substring(0, 5),
            quantity = addCartItemInput.quantity,
            product = productService.getProduct(addCartItemInput.productId),
            cart = getUserCart(addCartItemInput.userId)
        ).also { Database.cartItems.add(it) }

        return getUserCart(addCartItemInput.userId)
    }

    fun deleteCartItem(deleteCartItemInput: DeleteCartItemInput): Cart {
        Database.cartItems.removeIf { it.id == deleteCartItemInput.cartItemId }

        return getUserCart(deleteCartItemInput.userId)
    }

}